
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Essential Libraries For Science &#8212; Using Python for Science</title>
    
  <link href="_static/css/theme.css" rel="stylesheet" />
  <link href="_static/css/index.c5995385ac14fb8791e8eb36b4908be2.css" rel="stylesheet" />

    
  <link rel="stylesheet"
    href="_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="_static/sphinx-book-theme.e8e5499552300ddf5d7adccae7cc3b70.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="_static/js/index.1c5a1a01449ed65a7b51.js">

    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/togglebutton.js"></script>
    <script src="_static/clipboard.min.js"></script>
    <script src="_static/copybutton.js"></script>
    <script >var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="_static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <script async="async" src="https://unpkg.com/thebelab@latest/lib/index.js"></script>
    <script >
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="[🚧 Numpy 🚧]" href="numpy.html" />
    <link rel="prev" title="[🚧 Editors And Ides 🚧]" href="editors-and-ides.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="en" />
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="index.html">
      
      <img src="_static/logo.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">Using Python for Science</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main navigation">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="book-intro.html">
   Welcome to Using Python for Science
  </a>
 </li>
</ul>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="terminals-and-shells.html">
   [🚧 Terminals And Shells 🚧]
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="intro-to-environments.html">
   Introduction to Python environments
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="editors-and-ides.html">
   [🚧 Editors And Ides 🚧]
  </a>
 </li>
 <li class="toctree-l1 current active has-children">
  <a class="current reference internal" href="#">
   Essential Libraries For Science
  </a>
  <input checked="" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="numpy.html">
     [🚧 Numpy 🚧]
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="matplotlib.html">
     [🚧 Matplotlib 🚧]
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="pandas.html">
     [🚧 Pandas 🚧]
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="get-it-working.html">
   Get It Working
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="understanding-errors.html">
     Understanding Python errors
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="debugging.html">
     [🚧 Debugging 🚧]
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="get-it-right.html">
   [🚧 Get It Right 🚧]
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/>
  <label for="toctree-checkbox-3">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="testing.html">
     [🚧 Testing 🚧]
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="continuous-integration.html">
     [🚧 Continuous Integration 🚧]
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="reproducibility.html">
     [🚧 Reproducibility 🚧]
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="get-it-fast.html">
   [🚧 Get It Fast 🚧]
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/>
  <label for="toctree-checkbox-4">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="profiling.html">
     [🚧 Profiling 🚧]
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="cython.html">
     [🚧 Cython 🚧]
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="parallel-python.html">
     [🚧 Parallel Python 🚧]
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="numba.html">
     [🚧 Numba 🚧]
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="gpu-libraries.html">
     [🚧 Gpu Libraries 🚧]
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="ship-it.html">
   [🚧 Ship It 🚧]
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/>
  <label for="toctree-checkbox-5">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="packaging-for-pip.html">
     [🚧 Packaging For Pip 🚧]
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="packaging-for-conda.html">
     [🚧 Packaging For Conda 🚧]
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="epilogue.html">
   [🚧 Epilogue 🚧]
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        <a class="dropdown-buttons"
            href="_sources/essential-libraries-for-science.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download notebook file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="_sources/essential-libraries-for-science.md"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.md</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Launch interactive content"><i class="fas fa-rocket"></i></button>
    <div class="dropdown-buttons">
        
        <a class="binder-button" href="https://mybinder.org/v2/gh/executablebooks/jupyter-book/master?urlpath=tree/essential-libraries-for-science.md"><button type="button"
                class="btn btn-secondary topbarbtn" title="Launch Binder" data-toggle="tooltip"
                data-placement="left"><img class="binder-button-logo"
                    src="_static/images/logo_binder.svg"
                    alt="Interact on binder">Binder</button></a>
        
        
        
        
    </div>
</div>

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#core-libraries-numpy-scipy-pandas">
   Core libraries: numpy, scipy, pandas.
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#things-you-cannot-or-should-not-work-without">
   “Things you cannot (or should not) work without”
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#visualization-libraries">
   Visualization libraries
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#speed-ups">
   Speed ups
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#domain-specific">
   Domain specific
  </a>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="section" id="essential-libraries-for-science">
<h1>Essential Libraries For Science<a class="headerlink" href="#essential-libraries-for-science" title="Permalink to this headline">¶</a></h1>
<p>A core part of programming is having a vague understanding of what exists
out there in terms of functionality, tools, libraries, … You do not need to
know how to do a task, but that this task is possible using this or that
library. In this chapter, we provide a somewhat opiniated list of tools and
libraries that you should know about when writing code for science. We will
not teach <em>how</em> to use those libraries, as there are plenty of
available resources for that. Instead, we only provide a short
introduction of what those libraries <em>do</em>.</p>
<p>We have divided the list of libraries into 5 categories: (1) core libraries,
ie libraries that you cannot code for science without; (2) libraries and tools
you should not work without; (3) visualizationg and plotting libraries; (4)
things that will help you speed up your code; (5) domain specific libraries.</p>
<div class="section" id="core-libraries-numpy-scipy-pandas">
<h2>Core libraries: numpy, scipy, pandas.<a class="headerlink" href="#core-libraries-numpy-scipy-pandas" title="Permalink to this headline">¶</a></h2>
<p>About 95% of scientific Python code starts with the following lines:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">from</span> <span class="nn">scipy</span> <span class="kn">import</span> <span class="n">XXX</span>
</pre></div>
</div>
<ul class="simple">
<li><p><strong><code class="docutils literal notranslate"><span class="pre">numpy</span></code></strong>’s core functionality is to provide an easy-to-use and efficient
n-dimensional array data structure called <code class="docutils literal notranslate"><span class="pre">ndarray</span></code>. Any numerical analysis
code in Python heavily relies on the <code class="docutils literal notranslate"><span class="pre">ndarray</span></code> data structure. The main
difference with Python’s list is that arrays are almost always homogeneously
typed (all elements of the array are of the same type). <code class="docutils literal notranslate"><span class="pre">numpy</span></code> also
provides fast and easy operations such as sum, element-wise product, matrix
product, indexing, slicing, broadcasting, … Note that we have a full chapter
on how to use <code class="docutils literal notranslate"><span class="pre">numpy</span></code> (see <a class="reference internal" href="numpy.html"><span class="doc">[🚧 Numpy 🚧]</span></a> for details)!</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">scipy</span></code></strong> is the core toolbox for scientific and technical computing for
Python. It contains many submodules, some related to complex data structures
(for sparse matrices for example) or operations on <code class="docutils literal notranslate"><span class="pre">ndarray</span></code> (linear
algebra, fourier transforms, IO), others domain specific (signal processing,
image processing, clustering). View <code class="docutils literal notranslate"><span class="pre">scipy</span></code> as the Swiss army knife of
scientific computing in Python.</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">pandas</span></code></strong> is a software library for the manipulation of data frames (an
object to store tabular data, as you might find in Excel or relational
databases). This differs from NumPy arrays in being ‘column-oriented’: the
values in each column have to be of the same type, but one row can contain
many different types of objects.” Just as NumPy does for arrays,
<code class="docutils literal notranslate"><span class="pre">pandas</span></code> provides fast and easy operations on dataframes, such as pivoting,
merging, joining, data filtering, column-wise operations. We also have a
full chapter on how to use <code class="docutils literal notranslate"><span class="pre">pandas</span></code> (see <a class="reference internal" href="pandas.html"><span class="doc">[🚧 Pandas 🚧]</span></a> for details)!</p></li>
</ul>
</div>
<div class="section" id="things-you-cannot-or-should-not-work-without">
<h2>“Things you cannot (or should not) work without”<a class="headerlink" href="#things-you-cannot-or-should-not-work-without" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>A good text editor or an IDE (See <a class="reference internal" href="editors-and-ides.html"><span class="doc">[🚧 Editors And Ides 🚧]</span></a> for details)!</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">ipython</span></code> is a command shell for Python. Think <code class="docutils literal notranslate"><span class="pre">python</span></code>, but a hundred times
better! It has syntax highlighting, auto-completion, and many magic
functions (<code class="docutils literal notranslate"><span class="pre">%debug</span></code>, <code class="docutils literal notranslate"><span class="pre">%run</span></code>, etc) to ease the process of programming, data
exploration, and debugging.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">pdb</span></code>, <code class="docutils literal notranslate"><span class="pre">ipdb</span></code>, <code class="docutils literal notranslate"><span class="pre">pdb++</span></code>, and <code class="docutils literal notranslate"><span class="pre">pudb</span></code> are four python debuggers. Knowing how to use a
Python debugger is a must! If you don’t know how to use one, read
<a class="reference internal" href="debugging.html"><span class="doc">[🚧 Debugging 🚧]</span></a>) without further delay!</p></li>
<li><p>Whether you use an IDE or a text editor, configure it with a static linter
to check your code as you write it. There are different options available. A
widely used static linter is <code class="docutils literal notranslate"><span class="pre">flake8</span></code>: it combines three tools into one
(<code class="docutils literal notranslate"><span class="pre">pyflakes</span></code>, <code class="docutils literal notranslate"><span class="pre">pycodestyle</span></code>, and <code class="docutils literal notranslate"><span class="pre">mccabe</span></code>). Another popular one is <code class="docutils literal notranslate"><span class="pre">pylint</span></code>.</p></li>
</ul>
</div>
<div class="section" id="visualization-libraries">
<h2>Visualization libraries<a class="headerlink" href="#visualization-libraries" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p><strong><code class="docutils literal notranslate"><span class="pre">Matplotlib</span></code></strong> is the most widely used visualization library in Python.
While the API takes a while to get used to, there is nothing better to
create high quality scientific figures for publication! Literally everything
can be tweaked, tuned, and modified.</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">seaborn</span></code></strong> is a thin layer on top of Matplotlib: you get the flexibility
of <code class="docutils literal notranslate"><span class="pre">Matplotlib</span></code> with an easy-to-use interface.</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">bokeh</span></code></strong>, <strong><code class="docutils literal notranslate"><span class="pre">plotly</span></code></strong>, and <strong><code class="docutils literal notranslate"><span class="pre">altair</span></code></strong> are three excellent choices of
plotting libraries with interactive outputs. All of these play very well
with notebooks, such as myst-nb or jupyter notebooks. Note that <code class="docutils literal notranslate"><span class="pre">plotly</span></code> can
be used to create interactive tables as well. Because we can’t resist
showing off cool features, here’s a small code snippet we shamelessly stole
from the myst-nb documentation on creating an interactive table and plot.</p></li>
</ul>
<p>Here’s an interactive table with <code class="docutils literal notranslate"><span class="pre">plotly</span></code>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">plotly.express</span> <span class="k">as</span> <span class="nn">px</span>
<span class="n">data</span> <span class="o">=</span> <span class="n">px</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">iris</span><span class="p">()</span>
<span class="n">data</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>sepal_length</th>
      <th>sepal_width</th>
      <th>petal_length</th>
      <th>petal_width</th>
      <th>species</th>
      <th>species_id</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>5.1</td>
      <td>3.5</td>
      <td>1.4</td>
      <td>0.2</td>
      <td>setosa</td>
      <td>1</td>
    </tr>
    <tr>
      <th>1</th>
      <td>4.9</td>
      <td>3.0</td>
      <td>1.4</td>
      <td>0.2</td>
      <td>setosa</td>
      <td>1</td>
    </tr>
    <tr>
      <th>2</th>
      <td>4.7</td>
      <td>3.2</td>
      <td>1.3</td>
      <td>0.2</td>
      <td>setosa</td>
      <td>1</td>
    </tr>
    <tr>
      <th>3</th>
      <td>4.6</td>
      <td>3.1</td>
      <td>1.5</td>
      <td>0.2</td>
      <td>setosa</td>
      <td>1</td>
    </tr>
    <tr>
      <th>4</th>
      <td>5.0</td>
      <td>3.6</td>
      <td>1.4</td>
      <td>0.2</td>
      <td>setosa</td>
      <td>1</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>And now a small graph of the same data with <code class="docutils literal notranslate"><span class="pre">Altair</span></code>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">altair</span> <span class="k">as</span> <span class="nn">alt</span>
<span class="n">alt</span><span class="o">.</span><span class="n">Chart</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">)</span><span class="o">.</span><span class="n">mark_point</span><span class="p">()</span><span class="o">.</span><span class="n">encode</span><span class="p">(</span>
    <span class="n">x</span><span class="o">=</span><span class="s2">&quot;sepal_width&quot;</span><span class="p">,</span>
    <span class="n">y</span><span class="o">=</span><span class="s2">&quot;sepal_length&quot;</span><span class="p">,</span>
    <span class="n">color</span><span class="o">=</span><span class="s2">&quot;species&quot;</span><span class="p">,</span>
    <span class="n">size</span><span class="o">=</span><span class="s1">&#39;sepal_length&#39;</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
<div id="altair-viz-24f1d92f8cb54250a0972d4f06027ef3"></div>
<script type="text/javascript">
  (function(spec, embedOpt){
    let outputDiv = document.currentScript.previousElementSibling;
    if (outputDiv.id !== "altair-viz-24f1d92f8cb54250a0972d4f06027ef3") {
      outputDiv = document.getElementById("altair-viz-24f1d92f8cb54250a0972d4f06027ef3");
    }
    const paths = {
      "vega": "https://cdn.jsdelivr.net/npm//vega@5?noext",
      "vega-lib": "https://cdn.jsdelivr.net/npm//vega-lib?noext",
      "vega-lite": "https://cdn.jsdelivr.net/npm//vega-lite@4.8.1?noext",
      "vega-embed": "https://cdn.jsdelivr.net/npm//vega-embed@6?noext",
    };

    function loadScript(lib) {
      return new Promise(function(resolve, reject) {
        var s = document.createElement('script');
        s.src = paths[lib];
        s.async = true;
        s.onload = () => resolve(paths[lib]);
        s.onerror = () => reject(`Error loading script: ${paths[lib]}`);
        document.getElementsByTagName("head")[0].appendChild(s);
      });
    }

    function showError(err) {
      outputDiv.innerHTML = `<div class="error" style="color:red;">${err}</div>`;
      throw err;
    }

    function displayChart(vegaEmbed) {
      vegaEmbed(outputDiv, spec, embedOpt)
        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));
    }

    if(typeof define === "function" && define.amd) {
      requirejs.config({paths});
      require(["vega-embed"], displayChart, err => showError(`Error loading script: ${err.message}`));
    } else if (typeof vegaEmbed === "function") {
      displayChart(vegaEmbed);
    } else {
      loadScript("vega")
        .then(() => loadScript("vega-lite"))
        .then(() => loadScript("vega-embed"))
        .catch(showError)
        .then(() => displayChart(vegaEmbed));
    }
  })({"config": {"view": {"continuousWidth": 400, "continuousHeight": 300}}, "data": {"name": "data-98ae83e5d702045aa997eca40b770a99"}, "mark": "point", "encoding": {"color": {"type": "nominal", "field": "species"}, "size": {"type": "quantitative", "field": "sepal_length"}, "x": {"type": "quantitative", "field": "sepal_width"}, "y": {"type": "quantitative", "field": "sepal_length"}}, "$schema": "https://vega.github.io/schema/vega-lite/v4.8.1.json", "datasets": {"data-98ae83e5d702045aa997eca40b770a99": [{"sepal_length": 5.1, "sepal_width": 3.5, "petal_length": 1.4, "petal_width": 0.2, "species": "setosa", "species_id": 1}, {"sepal_length": 4.9, "sepal_width": 3.0, "petal_length": 1.4, "petal_width": 0.2, "species": "setosa", "species_id": 1}, {"sepal_length": 4.7, "sepal_width": 3.2, "petal_length": 1.3, "petal_width": 0.2, "species": "setosa", "species_id": 1}, {"sepal_length": 4.6, "sepal_width": 3.1, "petal_length": 1.5, "petal_width": 0.2, "species": "setosa", "species_id": 1}, {"sepal_length": 5.0, "sepal_width": 3.6, "petal_length": 1.4, "petal_width": 0.2, "species": "setosa", "species_id": 1}, {"sepal_length": 5.4, "sepal_width": 3.9, "petal_length": 1.7, "petal_width": 0.4, "species": "setosa", "species_id": 1}, {"sepal_length": 4.6, "sepal_width": 3.4, "petal_length": 1.4, "petal_width": 0.3, "species": "setosa", "species_id": 1}, {"sepal_length": 5.0, "sepal_width": 3.4, "petal_length": 1.5, "petal_width": 0.2, "species": "setosa", "species_id": 1}, {"sepal_length": 4.4, "sepal_width": 2.9, "petal_length": 1.4, "petal_width": 0.2, "species": "setosa", "species_id": 1}, {"sepal_length": 4.9, "sepal_width": 3.1, "petal_length": 1.5, "petal_width": 0.1, "species": "setosa", "species_id": 1}, {"sepal_length": 5.4, "sepal_width": 3.7, "petal_length": 1.5, "petal_width": 0.2, "species": "setosa", "species_id": 1}, {"sepal_length": 4.8, "sepal_width": 3.4, "petal_length": 1.6, "petal_width": 0.2, "species": "setosa", "species_id": 1}, {"sepal_length": 4.8, "sepal_width": 3.0, "petal_length": 1.4, "petal_width": 0.1, "species": "setosa", "species_id": 1}, {"sepal_length": 4.3, "sepal_width": 3.0, "petal_length": 1.1, "petal_width": 0.1, "species": "setosa", "species_id": 1}, {"sepal_length": 5.8, "sepal_width": 4.0, "petal_length": 1.2, "petal_width": 0.2, "species": "setosa", "species_id": 1}, {"sepal_length": 5.7, "sepal_width": 4.4, "petal_length": 1.5, "petal_width": 0.4, "species": "setosa", "species_id": 1}, {"sepal_length": 5.4, "sepal_width": 3.9, "petal_length": 1.3, "petal_width": 0.4, "species": "setosa", "species_id": 1}, {"sepal_length": 5.1, "sepal_width": 3.5, "petal_length": 1.4, "petal_width": 0.3, "species": "setosa", "species_id": 1}, {"sepal_length": 5.7, "sepal_width": 3.8, "petal_length": 1.7, "petal_width": 0.3, "species": "setosa", "species_id": 1}, {"sepal_length": 5.1, "sepal_width": 3.8, "petal_length": 1.5, "petal_width": 0.3, "species": "setosa", "species_id": 1}, {"sepal_length": 5.4, "sepal_width": 3.4, "petal_length": 1.7, "petal_width": 0.2, "species": "setosa", "species_id": 1}, {"sepal_length": 5.1, "sepal_width": 3.7, "petal_length": 1.5, "petal_width": 0.4, "species": "setosa", "species_id": 1}, {"sepal_length": 4.6, "sepal_width": 3.6, "petal_length": 1.0, "petal_width": 0.2, "species": "setosa", "species_id": 1}, {"sepal_length": 5.1, "sepal_width": 3.3, "petal_length": 1.7, "petal_width": 0.5, "species": "setosa", "species_id": 1}, {"sepal_length": 4.8, "sepal_width": 3.4, "petal_length": 1.9, "petal_width": 0.2, "species": "setosa", "species_id": 1}, {"sepal_length": 5.0, "sepal_width": 3.0, "petal_length": 1.6, "petal_width": 0.2, "species": "setosa", "species_id": 1}, {"sepal_length": 5.0, "sepal_width": 3.4, "petal_length": 1.6, "petal_width": 0.4, "species": "setosa", "species_id": 1}, {"sepal_length": 5.2, "sepal_width": 3.5, "petal_length": 1.5, "petal_width": 0.2, "species": "setosa", "species_id": 1}, {"sepal_length": 5.2, "sepal_width": 3.4, "petal_length": 1.4, "petal_width": 0.2, "species": "setosa", "species_id": 1}, {"sepal_length": 4.7, "sepal_width": 3.2, "petal_length": 1.6, "petal_width": 0.2, "species": "setosa", "species_id": 1}, {"sepal_length": 4.8, "sepal_width": 3.1, "petal_length": 1.6, "petal_width": 0.2, "species": "setosa", "species_id": 1}, {"sepal_length": 5.4, "sepal_width": 3.4, "petal_length": 1.5, "petal_width": 0.4, "species": "setosa", "species_id": 1}, {"sepal_length": 5.2, "sepal_width": 4.1, "petal_length": 1.5, "petal_width": 0.1, "species": "setosa", "species_id": 1}, {"sepal_length": 5.5, "sepal_width": 4.2, "petal_length": 1.4, "petal_width": 0.2, "species": "setosa", "species_id": 1}, {"sepal_length": 4.9, "sepal_width": 3.1, "petal_length": 1.5, "petal_width": 0.1, "species": "setosa", "species_id": 1}, {"sepal_length": 5.0, "sepal_width": 3.2, "petal_length": 1.2, "petal_width": 0.2, "species": "setosa", "species_id": 1}, {"sepal_length": 5.5, "sepal_width": 3.5, "petal_length": 1.3, "petal_width": 0.2, "species": "setosa", "species_id": 1}, {"sepal_length": 4.9, "sepal_width": 3.1, "petal_length": 1.5, "petal_width": 0.1, "species": "setosa", "species_id": 1}, {"sepal_length": 4.4, "sepal_width": 3.0, "petal_length": 1.3, "petal_width": 0.2, "species": "setosa", "species_id": 1}, {"sepal_length": 5.1, "sepal_width": 3.4, "petal_length": 1.5, "petal_width": 0.2, "species": "setosa", "species_id": 1}, {"sepal_length": 5.0, "sepal_width": 3.5, "petal_length": 1.3, "petal_width": 0.3, "species": "setosa", "species_id": 1}, {"sepal_length": 4.5, "sepal_width": 2.3, "petal_length": 1.3, "petal_width": 0.3, "species": "setosa", "species_id": 1}, {"sepal_length": 4.4, "sepal_width": 3.2, "petal_length": 1.3, "petal_width": 0.2, "species": "setosa", "species_id": 1}, {"sepal_length": 5.0, "sepal_width": 3.5, "petal_length": 1.6, "petal_width": 0.6, "species": "setosa", "species_id": 1}, {"sepal_length": 5.1, "sepal_width": 3.8, "petal_length": 1.9, "petal_width": 0.4, "species": "setosa", "species_id": 1}, {"sepal_length": 4.8, "sepal_width": 3.0, "petal_length": 1.4, "petal_width": 0.3, "species": "setosa", "species_id": 1}, {"sepal_length": 5.1, "sepal_width": 3.8, "petal_length": 1.6, "petal_width": 0.2, "species": "setosa", "species_id": 1}, {"sepal_length": 4.6, "sepal_width": 3.2, "petal_length": 1.4, "petal_width": 0.2, "species": "setosa", "species_id": 1}, {"sepal_length": 5.3, "sepal_width": 3.7, "petal_length": 1.5, "petal_width": 0.2, "species": "setosa", "species_id": 1}, {"sepal_length": 5.0, "sepal_width": 3.3, "petal_length": 1.4, "petal_width": 0.2, "species": "setosa", "species_id": 1}, {"sepal_length": 7.0, "sepal_width": 3.2, "petal_length": 4.7, "petal_width": 1.4, "species": "versicolor", "species_id": 2}, {"sepal_length": 6.4, "sepal_width": 3.2, "petal_length": 4.5, "petal_width": 1.5, "species": "versicolor", "species_id": 2}, {"sepal_length": 6.9, "sepal_width": 3.1, "petal_length": 4.9, "petal_width": 1.5, "species": "versicolor", "species_id": 2}, {"sepal_length": 5.5, "sepal_width": 2.3, "petal_length": 4.0, "petal_width": 1.3, "species": "versicolor", "species_id": 2}, {"sepal_length": 6.5, "sepal_width": 2.8, "petal_length": 4.6, "petal_width": 1.5, "species": "versicolor", "species_id": 2}, {"sepal_length": 5.7, "sepal_width": 2.8, "petal_length": 4.5, "petal_width": 1.3, "species": "versicolor", "species_id": 2}, {"sepal_length": 6.3, "sepal_width": 3.3, "petal_length": 4.7, "petal_width": 1.6, "species": "versicolor", "species_id": 2}, {"sepal_length": 4.9, "sepal_width": 2.4, "petal_length": 3.3, "petal_width": 1.0, "species": "versicolor", "species_id": 2}, {"sepal_length": 6.6, "sepal_width": 2.9, "petal_length": 4.6, "petal_width": 1.3, "species": "versicolor", "species_id": 2}, {"sepal_length": 5.2, "sepal_width": 2.7, "petal_length": 3.9, "petal_width": 1.4, "species": "versicolor", "species_id": 2}, {"sepal_length": 5.0, "sepal_width": 2.0, "petal_length": 3.5, "petal_width": 1.0, "species": "versicolor", "species_id": 2}, {"sepal_length": 5.9, "sepal_width": 3.0, "petal_length": 4.2, "petal_width": 1.5, "species": "versicolor", "species_id": 2}, {"sepal_length": 6.0, "sepal_width": 2.2, "petal_length": 4.0, "petal_width": 1.0, "species": "versicolor", "species_id": 2}, {"sepal_length": 6.1, "sepal_width": 2.9, "petal_length": 4.7, "petal_width": 1.4, "species": "versicolor", "species_id": 2}, {"sepal_length": 5.6, "sepal_width": 2.9, "petal_length": 3.6, "petal_width": 1.3, "species": "versicolor", "species_id": 2}, {"sepal_length": 6.7, "sepal_width": 3.1, "petal_length": 4.4, "petal_width": 1.4, "species": "versicolor", "species_id": 2}, {"sepal_length": 5.6, "sepal_width": 3.0, "petal_length": 4.5, "petal_width": 1.5, "species": "versicolor", "species_id": 2}, {"sepal_length": 5.8, "sepal_width": 2.7, "petal_length": 4.1, "petal_width": 1.0, "species": "versicolor", "species_id": 2}, {"sepal_length": 6.2, "sepal_width": 2.2, "petal_length": 4.5, "petal_width": 1.5, "species": "versicolor", "species_id": 2}, {"sepal_length": 5.6, "sepal_width": 2.5, "petal_length": 3.9, "petal_width": 1.1, "species": "versicolor", "species_id": 2}, {"sepal_length": 5.9, "sepal_width": 3.2, "petal_length": 4.8, "petal_width": 1.8, "species": "versicolor", "species_id": 2}, {"sepal_length": 6.1, "sepal_width": 2.8, "petal_length": 4.0, "petal_width": 1.3, "species": "versicolor", "species_id": 2}, {"sepal_length": 6.3, "sepal_width": 2.5, "petal_length": 4.9, "petal_width": 1.5, "species": "versicolor", "species_id": 2}, {"sepal_length": 6.1, "sepal_width": 2.8, "petal_length": 4.7, "petal_width": 1.2, "species": "versicolor", "species_id": 2}, {"sepal_length": 6.4, "sepal_width": 2.9, "petal_length": 4.3, "petal_width": 1.3, "species": "versicolor", "species_id": 2}, {"sepal_length": 6.6, "sepal_width": 3.0, "petal_length": 4.4, "petal_width": 1.4, "species": "versicolor", "species_id": 2}, {"sepal_length": 6.8, "sepal_width": 2.8, "petal_length": 4.8, "petal_width": 1.4, "species": "versicolor", "species_id": 2}, {"sepal_length": 6.7, "sepal_width": 3.0, "petal_length": 5.0, "petal_width": 1.7, "species": "versicolor", "species_id": 2}, {"sepal_length": 6.0, "sepal_width": 2.9, "petal_length": 4.5, "petal_width": 1.5, "species": "versicolor", "species_id": 2}, {"sepal_length": 5.7, "sepal_width": 2.6, "petal_length": 3.5, "petal_width": 1.0, "species": "versicolor", "species_id": 2}, {"sepal_length": 5.5, "sepal_width": 2.4, "petal_length": 3.8, "petal_width": 1.1, "species": "versicolor", "species_id": 2}, {"sepal_length": 5.5, "sepal_width": 2.4, "petal_length": 3.7, "petal_width": 1.0, "species": "versicolor", "species_id": 2}, {"sepal_length": 5.8, "sepal_width": 2.7, "petal_length": 3.9, "petal_width": 1.2, "species": "versicolor", "species_id": 2}, {"sepal_length": 6.0, "sepal_width": 2.7, "petal_length": 5.1, "petal_width": 1.6, "species": "versicolor", "species_id": 2}, {"sepal_length": 5.4, "sepal_width": 3.0, "petal_length": 4.5, "petal_width": 1.5, "species": "versicolor", "species_id": 2}, {"sepal_length": 6.0, "sepal_width": 3.4, "petal_length": 4.5, "petal_width": 1.6, "species": "versicolor", "species_id": 2}, {"sepal_length": 6.7, "sepal_width": 3.1, "petal_length": 4.7, "petal_width": 1.5, "species": "versicolor", "species_id": 2}, {"sepal_length": 6.3, "sepal_width": 2.3, "petal_length": 4.4, "petal_width": 1.3, "species": "versicolor", "species_id": 2}, {"sepal_length": 5.6, "sepal_width": 3.0, "petal_length": 4.1, "petal_width": 1.3, "species": "versicolor", "species_id": 2}, {"sepal_length": 5.5, "sepal_width": 2.5, "petal_length": 4.0, "petal_width": 1.3, "species": "versicolor", "species_id": 2}, {"sepal_length": 5.5, "sepal_width": 2.6, "petal_length": 4.4, "petal_width": 1.2, "species": "versicolor", "species_id": 2}, {"sepal_length": 6.1, "sepal_width": 3.0, "petal_length": 4.6, "petal_width": 1.4, "species": "versicolor", "species_id": 2}, {"sepal_length": 5.8, "sepal_width": 2.6, "petal_length": 4.0, "petal_width": 1.2, "species": "versicolor", "species_id": 2}, {"sepal_length": 5.0, "sepal_width": 2.3, "petal_length": 3.3, "petal_width": 1.0, "species": "versicolor", "species_id": 2}, {"sepal_length": 5.6, "sepal_width": 2.7, "petal_length": 4.2, "petal_width": 1.3, "species": "versicolor", "species_id": 2}, {"sepal_length": 5.7, "sepal_width": 3.0, "petal_length": 4.2, "petal_width": 1.2, "species": "versicolor", "species_id": 2}, {"sepal_length": 5.7, "sepal_width": 2.9, "petal_length": 4.2, "petal_width": 1.3, "species": "versicolor", "species_id": 2}, {"sepal_length": 6.2, "sepal_width": 2.9, "petal_length": 4.3, "petal_width": 1.3, "species": "versicolor", "species_id": 2}, {"sepal_length": 5.1, "sepal_width": 2.5, "petal_length": 3.0, "petal_width": 1.1, "species": "versicolor", "species_id": 2}, {"sepal_length": 5.7, "sepal_width": 2.8, "petal_length": 4.1, "petal_width": 1.3, "species": "versicolor", "species_id": 2}, {"sepal_length": 6.3, "sepal_width": 3.3, "petal_length": 6.0, "petal_width": 2.5, "species": "virginica", "species_id": 3}, {"sepal_length": 5.8, "sepal_width": 2.7, "petal_length": 5.1, "petal_width": 1.9, "species": "virginica", "species_id": 3}, {"sepal_length": 7.1, "sepal_width": 3.0, "petal_length": 5.9, "petal_width": 2.1, "species": "virginica", "species_id": 3}, {"sepal_length": 6.3, "sepal_width": 2.9, "petal_length": 5.6, "petal_width": 1.8, "species": "virginica", "species_id": 3}, {"sepal_length": 6.5, "sepal_width": 3.0, "petal_length": 5.8, "petal_width": 2.2, "species": "virginica", "species_id": 3}, {"sepal_length": 7.6, "sepal_width": 3.0, "petal_length": 6.6, "petal_width": 2.1, "species": "virginica", "species_id": 3}, {"sepal_length": 4.9, "sepal_width": 2.5, "petal_length": 4.5, "petal_width": 1.7, "species": "virginica", "species_id": 3}, {"sepal_length": 7.3, "sepal_width": 2.9, "petal_length": 6.3, "petal_width": 1.8, "species": "virginica", "species_id": 3}, {"sepal_length": 6.7, "sepal_width": 2.5, "petal_length": 5.8, "petal_width": 1.8, "species": "virginica", "species_id": 3}, {"sepal_length": 7.2, "sepal_width": 3.6, "petal_length": 6.1, "petal_width": 2.5, "species": "virginica", "species_id": 3}, {"sepal_length": 6.5, "sepal_width": 3.2, "petal_length": 5.1, "petal_width": 2.0, "species": "virginica", "species_id": 3}, {"sepal_length": 6.4, "sepal_width": 2.7, "petal_length": 5.3, "petal_width": 1.9, "species": "virginica", "species_id": 3}, {"sepal_length": 6.8, "sepal_width": 3.0, "petal_length": 5.5, "petal_width": 2.1, "species": "virginica", "species_id": 3}, {"sepal_length": 5.7, "sepal_width": 2.5, "petal_length": 5.0, "petal_width": 2.0, "species": "virginica", "species_id": 3}, {"sepal_length": 5.8, "sepal_width": 2.8, "petal_length": 5.1, "petal_width": 2.4, "species": "virginica", "species_id": 3}, {"sepal_length": 6.4, "sepal_width": 3.2, "petal_length": 5.3, "petal_width": 2.3, "species": "virginica", "species_id": 3}, {"sepal_length": 6.5, "sepal_width": 3.0, "petal_length": 5.5, "petal_width": 1.8, "species": "virginica", "species_id": 3}, {"sepal_length": 7.7, "sepal_width": 3.8, "petal_length": 6.7, "petal_width": 2.2, "species": "virginica", "species_id": 3}, {"sepal_length": 7.7, "sepal_width": 2.6, "petal_length": 6.9, "petal_width": 2.3, "species": "virginica", "species_id": 3}, {"sepal_length": 6.0, "sepal_width": 2.2, "petal_length": 5.0, "petal_width": 1.5, "species": "virginica", "species_id": 3}, {"sepal_length": 6.9, "sepal_width": 3.2, "petal_length": 5.7, "petal_width": 2.3, "species": "virginica", "species_id": 3}, {"sepal_length": 5.6, "sepal_width": 2.8, "petal_length": 4.9, "petal_width": 2.0, "species": "virginica", "species_id": 3}, {"sepal_length": 7.7, "sepal_width": 2.8, "petal_length": 6.7, "petal_width": 2.0, "species": "virginica", "species_id": 3}, {"sepal_length": 6.3, "sepal_width": 2.7, "petal_length": 4.9, "petal_width": 1.8, "species": "virginica", "species_id": 3}, {"sepal_length": 6.7, "sepal_width": 3.3, "petal_length": 5.7, "petal_width": 2.1, "species": "virginica", "species_id": 3}, {"sepal_length": 7.2, "sepal_width": 3.2, "petal_length": 6.0, "petal_width": 1.8, "species": "virginica", "species_id": 3}, {"sepal_length": 6.2, "sepal_width": 2.8, "petal_length": 4.8, "petal_width": 1.8, "species": "virginica", "species_id": 3}, {"sepal_length": 6.1, "sepal_width": 3.0, "petal_length": 4.9, "petal_width": 1.8, "species": "virginica", "species_id": 3}, {"sepal_length": 6.4, "sepal_width": 2.8, "petal_length": 5.6, "petal_width": 2.1, "species": "virginica", "species_id": 3}, {"sepal_length": 7.2, "sepal_width": 3.0, "petal_length": 5.8, "petal_width": 1.6, "species": "virginica", "species_id": 3}, {"sepal_length": 7.4, "sepal_width": 2.8, "petal_length": 6.1, "petal_width": 1.9, "species": "virginica", "species_id": 3}, {"sepal_length": 7.9, "sepal_width": 3.8, "petal_length": 6.4, "petal_width": 2.0, "species": "virginica", "species_id": 3}, {"sepal_length": 6.4, "sepal_width": 2.8, "petal_length": 5.6, "petal_width": 2.2, "species": "virginica", "species_id": 3}, {"sepal_length": 6.3, "sepal_width": 2.8, "petal_length": 5.1, "petal_width": 1.5, "species": "virginica", "species_id": 3}, {"sepal_length": 6.1, "sepal_width": 2.6, "petal_length": 5.6, "petal_width": 1.4, "species": "virginica", "species_id": 3}, {"sepal_length": 7.7, "sepal_width": 3.0, "petal_length": 6.1, "petal_width": 2.3, "species": "virginica", "species_id": 3}, {"sepal_length": 6.3, "sepal_width": 3.4, "petal_length": 5.6, "petal_width": 2.4, "species": "virginica", "species_id": 3}, {"sepal_length": 6.4, "sepal_width": 3.1, "petal_length": 5.5, "petal_width": 1.8, "species": "virginica", "species_id": 3}, {"sepal_length": 6.0, "sepal_width": 3.0, "petal_length": 4.8, "petal_width": 1.8, "species": "virginica", "species_id": 3}, {"sepal_length": 6.9, "sepal_width": 3.1, "petal_length": 5.4, "petal_width": 2.1, "species": "virginica", "species_id": 3}, {"sepal_length": 6.7, "sepal_width": 3.1, "petal_length": 5.6, "petal_width": 2.4, "species": "virginica", "species_id": 3}, {"sepal_length": 6.9, "sepal_width": 3.1, "petal_length": 5.1, "petal_width": 2.3, "species": "virginica", "species_id": 3}, {"sepal_length": 5.8, "sepal_width": 2.7, "petal_length": 5.1, "petal_width": 1.9, "species": "virginica", "species_id": 3}, {"sepal_length": 6.8, "sepal_width": 3.2, "petal_length": 5.9, "petal_width": 2.3, "species": "virginica", "species_id": 3}, {"sepal_length": 6.7, "sepal_width": 3.3, "petal_length": 5.7, "petal_width": 2.5, "species": "virginica", "species_id": 3}, {"sepal_length": 6.7, "sepal_width": 3.0, "petal_length": 5.2, "petal_width": 2.3, "species": "virginica", "species_id": 3}, {"sepal_length": 6.3, "sepal_width": 2.5, "petal_length": 5.0, "petal_width": 1.9, "species": "virginica", "species_id": 3}, {"sepal_length": 6.5, "sepal_width": 3.0, "petal_length": 5.2, "petal_width": 2.0, "species": "virginica", "species_id": 3}, {"sepal_length": 6.2, "sepal_width": 3.4, "petal_length": 5.4, "petal_width": 2.3, "species": "virginica", "species_id": 3}, {"sepal_length": 5.9, "sepal_width": 3.0, "petal_length": 5.1, "petal_width": 1.8, "species": "virginica", "species_id": 3}]}}, {"mode": "vega-lite"});
</script></div></div>
</div>
</div>
<div class="section" id="speed-ups">
<h2>Speed ups<a class="headerlink" href="#speed-ups" title="Permalink to this headline">¶</a></h2>
<p>Once you have a performing piece of code, it may come the time where you have
to speed your code up. Keep in mind that “Premature optimization is the root
of all evil,” so don’t look too closely at the list of packages in this
section unless you are sure that (1) your code works; (2) you really need to
speed things up.</p>
<ul class="simple">
<li><p>The first step of speeding up your code is to profile it: no need to spend
time optimizing the data analysis part if 90% of the time is spend on
loading the data. In order to do this, profile your code! There are many
profilers out there, but our favorite ones are <strong><code class="docutils literal notranslate"><span class="pre">line_profiler</span></code></strong> and
<strong><code class="docutils literal notranslate"><span class="pre">kernprof</span></code></strong>. <code class="docutils literal notranslate"><span class="pre">line_profiler</span></code> allows to do line-by-line profiling of
functions, while <code class="docutils literal notranslate"><span class="pre">kernprof</span></code> is a convenient script for running a
line-profiler (either <code class="docutils literal notranslate"><span class="pre">line_profiler</span></code> or <code class="docutils literal notranslate"><span class="pre">cProfile</span></code>, which is part of Python
standard library). If you are interested in profiling memory, we recommend
<strong><code class="docutils literal notranslate"><span class="pre">memprofiler</span></code></strong>. We invite you to read <a class="reference internal" href="profiling.html"><span class="doc">[🚧 Profiling 🚧]</span></a> if you
are at this step of the speed-up process!</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">cython</span></code></strong> is a programming language. It is a <em>superset</em> of Python, which
means that any valid Python is <em>also</em> valid Cython. However, it supports
<em>additional</em> syntax that allows it to be closer to the C language. Cython
thus sits between Python and C and serves two purposes: (1) easily interface
Python code and C code; (2) a Python/Cython to C compiler. The goal of
Cython is to give C-like performance with the ease of readability and
flexibility of Python. See <a class="reference internal" href="cython.html"><span class="doc">[🚧 Cython 🚧]</span></a> for more details!</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">numba</span></code></strong> is a JIT compiler for Python, as well as a parallelization
toolkit. It uses LLVM to generate machine code from Python code.
Specifically designed for scientific purposes, it can yield 100x speed ups
to a function by simply adding a decorator to it. We also give more details
about using it in <a class="reference internal" href="numba.html"><span class="doc">[🚧 Numba 🚧]</span></a>.</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">joblib</span></code></strong> provides a set of tools to create lightweight pipelining in
Python. Precisely, it has three main features: (1) fast disk-caching
capabilities; (2) embarassingly parallel helper functions; (3) and fast
compressed persistence for large data. The power of this library comes from
how lightweight it is and non-intrusive. You can use many of <code class="docutils literal notranslate"><span class="pre">joblib</span></code>’s
feature with barely any changes to the original code.</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">dask</span></code></strong> is a very powerful, parallelization libraries for data analytics.
It integrates very nicely with existing projects, such as <code class="docutils literal notranslate"><span class="pre">numpy</span></code>, <code class="docutils literal notranslate"><span class="pre">pandas</span></code>,
and <code class="docutils literal notranslate"><span class="pre">scikit-learn</span></code>. We won’t go into much details here, as we once again have
a full chapter on this library! See <a class="reference internal" href="parallel-python.html"><span class="doc">[🚧 Parallel Python 🚧]</span></a> for more details.</p></li>
</ul>
</div>
<div class="section" id="domain-specific">
<h2>Domain specific<a class="headerlink" href="#domain-specific" title="Permalink to this headline">¶</a></h2>
<p>The problem when it comes to domain specific libraries is that, well… They are
domain specific. As a result, only specialists of each domain know which
library to use when. Yet, unlike some communities (such as R), there are still
widely used and maintained packages for some domain. For example, the go-to
library to do astronomy data analysis in Python is <code class="docutils literal notranslate"><span class="pre">astropy</span></code>: none of us,
authors, are astronomers, yet we know that it is the go-to library (but,
please, don’t ask us more about <code class="docutils literal notranslate"><span class="pre">astropy</span></code>…).</p>
<p>Here, we provide a list of packages, with a small description, and sometimes a
bit more detail about when to use which package.</p>
<ul class="simple">
<li><p><strong><code class="docutils literal notranslate"><span class="pre">statsmodel</span></code></strong>: if you need to do standard statistics in Python, and
<code class="docutils literal notranslate"><span class="pre">scipy.stats</span></code> doesn’t cover it, there’s a good chance it can be done with
<code class="docutils literal notranslate"><span class="pre">statsmodel</span></code>.  It covers generalized least squares, quantile regression,
linear mixed-effects model (no, you do not have to use R to do those!!).</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">scikit-learn</span></code></strong> is the go-to libary for machine learning that isn’t deep:
SVM, regression models, random forests, cross-validation, and many metrics
are included in this package. But if you need a p-value out of your model,
you should check out <code class="docutils literal notranslate"><span class="pre">statsmodels</span></code> and not <code class="docutils literal notranslate"><span class="pre">scikit-learn</span></code>. If you want to do
deep learning, then check out our next element in the list.</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">tensorflow</span></code></strong> is the go-to library for doing deep-learning &amp; neural networks
in Python.</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">scikit-image</span></code></strong> complements <code class="docutils literal notranslate"><span class="pre">scipy.ndimage</span></code> for image processing algorithms.</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">networkx</span></code></strong> is meant for network analysis in Python</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">biopython</span></code></strong> provides support for biological data analysis in Python
(sequencing, alignment, population genetics algorithm, and structure
bioinformatrics).<a class="footnote-reference brackets" href="#de-analysis" id="id1">1</a></p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">astropy</span></code></strong> is, as mentioned, a toolbox for astronomy data analysis.</p></li>
</ul>
<hr class="footnotes docutils" />
<dl class="footnote brackets">
<dt class="label" id="de-analysis"><span class="brackets"><a class="fn-backref" href="#id1">1</a></span></dt>
<dd><p>Unfortunately, if you are looking for a tool to do
differential expression analysis, nothing much exists and the standard is
still to use R and bioconductor…</p>
</dd>
</dl>
</div>
<div class="toctree-wrapper compound">
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
        
        <div class='prev-next-bottom'>
            
    <a class='left-prev' id="prev-link" href="editors-and-ides.html" title="previous page">[🚧 Editors And Ides 🚧]</a>
    <a class='right-next' id="next-link" href="numpy.html" title="next page">[🚧 Numpy 🚧]</a>

        </div>
        
        </div>
    </div>
    <footer class="footer mt-5 mt-md-0">
    <div class="container">
      <p>
        
          By Juan Nunez-Iglesias & Nelle Varoquaux<br/>
        
            &copy; Copyright 2021.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="_static/js/index.1c5a1a01449ed65a7b51.js"></script>

  
  </body>
</html>